train.py

```diff
@@ -30,79 +33,80 @@ def normalize(data,rescaleInputToMax=False):
+    )
 
-  plt.plot(history.history['loss'])
-  plt.plot(history.history['val_loss'])
-  plt.title('Model loss %s'%name)
-  plt.ylabel('Loss')
-  plt.xlabel('Epoch')
-  plt.legend(['Train', 'Test'], loc='upper right')
-  plt.savefig("history_%s.png"%name)
-  #plt.show()
+    plt.plot(history.history['loss'])
+    plt.plot(history.history['val_loss'])
+    plt.title('Model loss %s'%name)
+    plt.ylabel('Loss')
+    plt.xlabel('Epoch')
+    plt.legend(['Train', 'Test'], loc='upper right')
+    plt.savefig("history_%s.png"%name)
+    #plt.show()
 
-  save_models(autoencoder,name)
+    save_models(autoencoder,name)
 
-  return history
+    return history
 
 def save_models(autoencoder, name):
-  json_string = autoencoder.to_json()
-  with open('./%s.json'%name,'w') as f:
-      f.write(json_string)
-  encoder = autoencoder.get_layer("encoder")
-  json_string = encoder.to_json()
-  with open('./%s.json'%("encoder_"+name),'w') as f:
-      f.write(json_string)
-  decoder = autoencoder.get_layer("decoder")
-  json_string = decoder.to_json()
-  with open('./%s.json'%("decoder_"+name),'w') as f:
-      f.write(json_string)
-  autoencoder.save_weights('%s.hdf5'%name)
-  encoder.save_weights('%s.hdf5'%("encoder_"+name))
-  decoder.save_weights('%s.hdf5'%("decoder_"+name))
-  return
-  
+    json_string = autoencoder.to_json()
+    with open('./%s.json'%name,'w') as f:
+        f.write(json_string)
+        encoder = autoencoder.get_layer("encoder")
+        json_string = encoder.to_json()
+        with open('./%s.json'%("encoder_"+name),'w') as f:
+            f.write(json_string)
+            decoder = autoencoder.get_layer("decoder")
+            json_string = decoder.to_json()
+            with open('./%s.json'%("decoder_"+name),'w') as f:
+                f.write(json_string)
+                autoencoder.save_weights('%s.hdf5'%name)
+                encoder.save_weights('%s.hdf5'%("encoder_"+name))
+                decoder.save_weights('%s.hdf5'%("decoder_"+name))
+                return
+                
 
 def predict(x,autoencoder,encoder,reshape=True):
-  decoded_Q = autoencoder.predict(x)
-  encoded_Q = encoder.predict(x)
- 
-  #need reshape for CNN layers  
-  if reshape :
-    decoded_Q = np.reshape(decoded_Q,(len(decoded_Q),12,4))
-    encoded_shape = encoded_Q.shape
-    encoded_Q = np.reshape(encoded_Q,(len(encoded_Q),encoded_shape[3],encoded_shape[1]))
-  return decoded_Q, encoded_Q
+    decoded_Q = autoencoder.predict(x)
+    encoded_Q = encoder.predict(x)
+    
+    #need reshape for CNN layers  
+    if reshape :
+        decoded_Q = np.reshape(decoded_Q,(len(decoded_Q),12,4))
+        encoded_shape = encoded_Q.shape
+        encoded_Q = np.reshape(encoded_Q,(len(encoded_Q),encoded_shape[3],encoded_shape[1]))
+
+    return decoded_Q, encoded_Q
 
 ### cross correlation of input/output 
 def cross_corr(x,y):
     cov = np.cov(x.flatten(),y.flatten())
-    std = np.sqrt(np.diag(cov))
+    std = np.sqrt(np.diag(cov)+1e-10)
```

Comment:
this should be without the 1e-10?

Reply:
do we need to protect against zero division? that was the idea here

Reply:
though we could do it in a smarter way

Reply:
I see, i guess it would protect cases like 1 epoch test run for completing. We could leave that for now